akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala

```diff
@@ -218,17 +216,25 @@ object Actor {
    * It is possible to specify different supervisor strategies, such as restart,
    * resume, backoff.
    *
+   * Note that only [[scala.util.control.NonFatal]] throwables will trigger the supervision strategy.
+   *
    * Example:
    * {{{
    * val dbConnector: Behavior[DbCommand] = ...
-   * val dbRestarts = Restarter[DbException]().wrap(dbConnector)
+   * val dbRestarts = supervise(dbConnector).onFailure(SupervisorStrategy.resume) // handle all NonFatal exceptions
+   * val dbSpecificRestarts = supervise(dbConnector).onFailure[IndexOutOfBoundsException](SupervisorStrategy.resume) // handle all NonFatal exceptions
    * }}}
    */
-  def restarter[Thr <: Throwable: ClassTag](strategy: SupervisorStrategy = SupervisorStrategy.restart): Restarter[Thr] =
-    new Restarter(implicitly, strategy)
-
-  final class Restarter[Thr <: Throwable: ClassTag](c: ClassTag[Thr], strategy: SupervisorStrategy) {
-    def wrap[T](b: Behavior[T]): Behavior[T] = akka.typed.internal.Restarter(Behavior.validateAsInitial(b), strategy)(c)
+  def supervise[T](wrapped: Behavior[T]): Supervise[T] =
+    new Supervise(wrapped)
+
+  final class Supervise[T](val wrapped: Behavior[T]) extends AnyVal {
+    /** Specify the [[SupervisorStrategy]] to be invoked when the wrapped behaior throws. */
+    def onFailure[Thr <: Throwable: ClassTag](strategy: SupervisorStrategy): Behavior[T] = {
+      val tag = implicitly[ClassTag[Thr]]
+      val effectiveTag = if (tag == ClassTag(classOf[Nothing])) ClassTag(classOf[Throwable]) else tag
```

Comment:
This is sneaky :-/
I'm open to any other ideas how to solve this. The problem is: sadly when someone would write `.onFailure(blabla)` (missing `[...]`), then Nothing would get inferred so we're attempting to match if a thing is Nothing, and not Throwable in the internals then.

I tried some type level hackery sadly that's not working : https://stackoverflow.com/questions/27912014/enforce-a-lower-bounds-that-excludes-a-super-type